core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+
+    private boolean needsRefresh(CacheEntry<T> cacheEntry) {
+        boolean refresh = false;
+        final T cache = cacheEntry.value;
+        try {
+            refresh = refreshIf != null && refreshIf.test(cache);
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from refreshIf.test()", e);
+        }
+
+        return refresh;
+    }
+
+    private void handleException(Throwable cause, @Nullable T cache,
+                                 CompletableFuture<CacheEntry<T>> future) {
+        if (exceptionHandler == null) {
+            future.completeExceptionally(cause);
+            return;
+        }
+
+        try {
+            final CompletableFuture<T> fallback = exceptionHandler.apply(cause, cache);
+            if (fallback == null) {
+                future.completeExceptionally(cause);
+                return;
+            }
+
+            fallback.handle((value, cause0) -> {
+                if (cause0 != null) {
+                    logger.warn("Failed to load a new value from exceptionHandler: {}. " +
+                                "cache: {}", exceptionHandler, cache, cause0);
+                    future.completeExceptionally(cause0);
+                } else {
+                    future.complete(new CacheEntry<>(value));
+                }
+                return null;
+            });
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from exceptionHandler.apply()", e);
+            future.completeExceptionally(cause);
+        }
+    }
+
+    private boolean isValid(@Nullable CacheEntry<T> cacheEntry) {
+        if (cacheEntry == null) {
+            return false;
+        }
+
+        if (expireAfterLoadNanos > 0) {
+            final long elapsed = System.nanoTime() - cacheEntry.cachedAt;
+            if (elapsed >= expireAfterLoadNanos) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("The cached value expired after {} ms. cache: {}",
+                                 TimeUnit.NANOSECONDS.toMillis(expireAfterLoadNanos), cacheEntry.value);
+                }
+                return false;
+            }
+        }
+
+        try {
+            if (expireIf != null && expireIf.test(cacheEntry.value)) {
+                logger.debug("The cached value expired due to 'expireIf' condition. cache: {}",
+                             cacheEntry.value);
+                return false;
+            }
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from expireIf.test()", e);
+        }
+
+        return true;
+    }
+
+    private static class CacheEntry<T> {
+
+        private final T value;
+        private final long cachedAt = System.nanoTime();
+
+        CacheEntry(T value) {
+            requireNonNull(value, "value");
+            this.value = value;
+        }
+    }
+
+    private static class RefreshingFuture<U> extends CompletableFuture<CacheEntry<U>> {
```

Comment:
- We should not allow a user to change the state of this future via `complete*()` and `obtrude*()`.
- Can we create a new future when refreshing a new value, rather than using `obtrude*()`? I wouldn't risk using `obtrude*()` here because:
  - `obtrude*()` is designed by error recovery actions, which is not the case here
  - The future returned by `get()` can be reused by a user later; and
  - We can do without it.

Reply:
> The future returned by get() can be reused by a user later; and

A `CompletableFuture` is created and `RefreshingFuture` is not exposed to users.
https://github.com/line/armeria/pull/5590/files#diff-7233778a06886166fb9453feeb1062411a58488bc6f37cbe3aabe35999f01b7eR71

Reply:
> Can we create a new future when refreshing a new value, rather than using obtrude*()?

I came up with `RefreshingFuture` to simplify the implementation.
- Without `RefreshingFuture` we would have to manage two futures and the implementation would be more complicated.
- If there is no refresh condition, it is simple. When it expires, you can replace the main future.
- In the case of refresh, you must create a separate future while maintaining the main future. This is because the existing value may still be valid.
  - If the existing value expires in a refresh situation, a refreshing future must be returned and the main future must not be returned.
  - When the refresh future is completed, the main future needs to be updated again.  

I wanted to eliminate the two states and manage this with a single future.

If you think that using `obtrude*` API is inappropriate, it would be a good idea to create something else that can manage the refreshing and cached state with a single object.

Reply:
> Can we create a new future when refreshing a new value, rather than using obtrude*()
```java
// if needRefresh or expired
future = new RefreshingFuture(cacheEntry) // create new future!
load(cache, future);
return future;
```
On refreshing, we load new value with newly created `CompletableFuture` ~!

Reply:
I'm not convinced that it justifies the current usage of `obtrude`, which is beyond its original intention.